### DISCUSSION TITLE
Referring to an aliased table by it's dynamic name

### DISCUSSION THREAD

**Discussion by connebs** [sqlalchemy/sqlalchemy#6006] [Usage Questions] [answered]
Hi, I have a somewhat particular use-case and am having trouble figuring out the cleanest way to accomplish it.

In summary, I have this:
```py
p1 = aliased(Piece, name='p1')
query = select(
    func.json_agg(func.to_json(sa.literal_column(inspect(p1).name)))
).select_from(p1)
```

Where `Piece` is an ORM model. I need to have this model aliased for the purposes of a broader query, hence the `p1`. However, I also want to refer to this alias within my `select()` statement. The above is the best way I've found to do it so far, but it's a bit clunky. It requires me to:

- use a `name=` kwarg in `aliased()`
- use `sqlalchemy.inspect().name` to get that name back out.

But I don't actually care what the name is, and using something like `inspect()` seems like over-kill. I could also do this of course:

```py
p1 = aliased(Piece, name='p1')
query = select(
    func.json_agg(func.to_json(sa.literal_column('p1')))
).select_from(p1)
```

But that requires me ensuring that I've typed two strings correctly, which I think is best avoided.

What I would ideally like is something like this:

```py
p1 = aliased(Piece)
query = select(
    func.json_agg(func.to_json(sa.literal_column(p1.name)))
).select_from(p1)
```

**Comment by zzzeek** [comment, marked as answer]
are you using 1.4?  you can just do this:

```
p1 = aliased(Piece)

query = select(
    func.json_agg(func.to_json(inspect(p1).selectable.table_valued()))
).select_from(p1)

```

the inspect() part should be fixed so that you can call table_valued(p1), we don't have that yet.

**Reply by zzzeek** [reply]
or this:
```

p1 = Piece.__table__.alias()

query = select(
    func.json_agg(func.to_json(p1.table_valued()))
).select_from(p1)

```

**Reply by zzzeek** [reply]
this was all added as part of #3566.   you're using the new select() so you're on 1.4, full docs for the new stuff in terms of PostgreSQL's features is at https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#postgresql-table-valued-overview

**Reply by connebs** [reply]
Yep, on 1.4.

Thanks for the suggestions, I think `inspect(p1).selectable.table_valued()` is about what I was looking for. Though yes, `table_valued(p1)` would certainly be even better.
